source/game/system/RaceManager.cc

```diff
@@ -265,7 +286,64 @@ void RaceManager::Player::decrementLap() {
 /// @addr{0x805349B8}
 void RaceManager::Player::incrementLap() {
     m_maxKcp = 0;
-    ++m_currentLap;
+    if (++m_currentLap <= m_maxLap) {
+        return;
+    }
+
+    const auto *kart = Kart::KartObjectManager::Instance()->object(0);
+    u16 addMs = CourseMap::Instance()->getCheckPointEntryOffsetMs(m_checkpointId, kart->pos(),
+            kart->prevPos());
+
+    const Timer &currentTimer = RaceManager::Instance()->timerManager().currentTimer();
+
+    Timer timer;
+    timer.mil = static_cast<u16>(static_cast<f32>(addMs) + static_cast<f32>(currentTimer.mil));
+
+    if (timer.mil > 999) {
+        // The additional milliseconds caused an overflow
+        u8 seconds = currentTimer.sec;
+        timer.mil -= 1000;
+        timer.sec = seconds + 1;
+
+        u16 addMin = 0;
+        if (timer.sec > 59) {
+            timer.sec -= 60;
+            addMin = 1;
+        }
+
+        timer.min = addMin + currentTimer.min;
+        if (timer.min > 999) {
+            timer.min = 999;
+            timer.sec = 59;
+            timer.mil = 999;
+        }
+    } else {
+        timer.sec = currentTimer.sec;
+        timer.min = currentTimer.min;
+    }
+
+    timer.valid = true;
+    Timer &lapTimer = m_lapTimers[m_maxLap - 1];
```

Comment:
This will cause out of bounds on some CTs. Worry about it now or later?

Reply:
We currently cap the max lap as the base game does, so this won't be out of bounds for a while.

Reply:
Actually, this can cause out of bounds writes in the base game, as you found yourself. https://www.youtube.com/watch?v=O1BHac3pw6w&t=793s

The timers are dynamically allocated, and you're clobbering data in an entirely separate allocation. This apparently includes a flag or something similar with the model.

For now, we will assert that maxLap - 1 is less than the size of the lap timer array, but we will need to handle it more gracefully at some point.